File: 0981-time-based-key-value-store/0981-time-based-key-value-store.rb
class TimeMap
    def initialize()
        @map = Hash.new { |h, k| h[k] = [] }
    end


=begin
    :type key: String
    :type value: String
    :type timestamp: Integer
    :rtype: Void
=end
    def set(key, value, timestamp)
        @map[key] << [timestamp, value]
    end


=begin
    :type key: String
    :type timestamp: Integer
    :rtype: String
=end
    def get(key, timestamp)
        return "" if !@map.has_key?(key)
        
        stack = @map[key]
        low = 0
        high = stack.length - 1
        result = ""
        
        while low <= high
            mid = (high + low) / 2
            if timestamp == stack[mid][0]
                return stack[mid][1]
            elsif timestamp >= stack[mid][0]
                result = stack[mid][1]
                low = mid + 1
            else
                high = mid - 1
            end
        end

        return result
    end


end

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap.new()
# obj.set(key, value, timestamp)
# param_2 = obj.get(key, timestamp)

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.js

// can use a simple JavaScript object/hash for the key-value store
// the value is an array of [timestamp, value] pairs

// assume values are set in timestamp order (sorted by timestamp ascending)

// so we can use binary search when looking for a value by timestamp
// typical b-search searching for the target timestamp, or the highest timestamp < target
// go until low > high, where high is the highest possible timestamp
// return the mid value during search, if exact timestamp was found
// if final high is out of bounds, meaning no such timestamp exists, then return ""
// otherwise return the value at that high index

var TimeMap = function() {
    this.store = {};
};

/** 
 * @param {string} key 
 * @param {string} value 
 * @param {number} timestamp
 * @return {void}
 */
TimeMap.prototype.set = function(key, value, timestamp) {
    key in this.store ? this.store[key].push([timestamp, value])
                      : this.store[key] = [[timestamp, value]];
};

/** 
 * @param {string} key 
 * @param {number} timestamp
 * @return {string}
 */
TimeMap.prototype.get = function(key, timestamp) {
    if (!(key in this.store)) return "";

    const values = this.store[key];
    let low = 0;
    let high = values.length - 1;

    while (low <= high) {
        const mid = Math.floor((high + low) / 2);
        const [midTime, midValue] = values[mid];

        if (timestamp === midTime) {
            return midValue;
        } else if (timestamp > midTime) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return values[high] ?  values[high][1] : "";
};

/** 
 * Your TimeMap object will be instantiated and called as such:
 * var obj = new TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */